public/app/features/alerting/unified/components/rule-editor/FolderPickerWithAdd.tsx

```diff
@@ -0,0 +1,279 @@
+  };
+
+  useEffect(() => {
+    // if this is not the same as our initial value notify parent
+    if (folder && folder.value !== initialFolderId) {
+      onChange({ id: folder.value!, title: folder.label! });
+    }
+  }, [folder, initialFolderId, onChange]);
+
+  // initial values for dropdown
+  useAsync(async () => {
+    if (skipInitialLoad) {
+      const folder = await getInitialValues({
+        getFolder: getFolderById,
+        folderId: initialFolderId,
+        folderName: initialTitle,
+      });
+      setFolder(folder);
+    }
+
+    await loadInitialValue();
+  }, [skipInitialLoad, initialFolderId, initialTitle]);
+
+  useEffect(() => {
+    if (folder && folder.id === VALUE_FOR_ADD) {
+      setIsCustom(true);
+      setFolderLabel('Press enter to create and confirm the new folder.');
+    }
+  }, [folder, setFolderLabel]);
+
+  const onFolderChange = useCallback(
+    (newFolder: SelectableValue<number>, actionMeta: ActionMeta) => {
+      const value = newFolder.value;
+      if (value === VALUE_FOR_ADD) {
+        setFolder({ id: VALUE_FOR_ADD, title: newFolder?.title?.split('/').join('') || '' }); //slashes not allowed
+      } else {
+        if (!newFolder) {
+          newFolder = { value: 0, label: rootName };
+        }
+
+        if (actionMeta.action === 'clear' && onClear) {
+          onClear();
+          return;
+        }
+
+        setFolder(newFolder);
+        onChange({ id: newFolder.value!, title: newFolder.label! });
+      }
+    },
+    [onChange, onClear, rootName]
+  );
+
+  const createNewFolder = useCallback(
+    async (folderName: string) => {
+      const newFolder = await createFolder({ title: folderName });
+      let folder: SelectableValue<number> = { value: -1, label: 'Not created' };
+
+      if (newFolder.id > -1) {
+        appEvents.emit(AppEvents.alertSuccess, ['Folder Created', 'OK']);
+        folder = { value: newFolder.id, label: newFolder.title };
+
+        setFolder(newFolder);
+        onFolderChange(folder, { action: 'create-option', option: folder });
+      } else {
+        appEvents.emit(AppEvents.alertError, ['Folder could not be created']);
+      }
+
+      return folder;
+    },
+    [onFolderChange]
+  );
+
+  const onKeyDown = useCallback(
+    (event: React.KeyboardEvent) => {
+      switch (event.key) {
+        case 'Enter': {
+          createNewFolder(folder?.title!);
+          setIsCustom(false);
+          setFolderLabel(initialFolderLabel);
+          break;
+        }
+        case 'Escape': {
+          setFolder({ value: 0, label: rootName });
+          setIsCustom(false);
+          setFolderLabel(initialFolderLabel);
+        }
+      }
+    },
+    [createNewFolder, folder, rootName, setFolderLabel]
+  );
+
+  const onNewFolderChange = (e: FormEvent<HTMLInputElement>) => {
+    const value = e.currentTarget.value;
+    const withoutSlashes = value.split('/').join('');
+    setFolder({ id: -1, title: withoutSlashes });
+  };
+
+  if (isCustom) {
```

Comment:
`react-select` supports `AsyncCreatable`. Maybe we could use that instead of handling it on our own?
Take a look at https://github.com/grafana/grafana/blob/55b1d11aceb1081d62fc01e20f38f0013866fa43/packages/grafana-ui/src/components/Select/SelectBase.tsx#L260-L267

Reply:
So far, we are using Asynccreatable : 
```
 if (allowCustomValue) {
      SelectComponent = Creatable;
      creatableOptions.formatCreateLabel = formatCreateLabel ?? ((input: string) => input);
    }
```
that comes from : `import Creatable from 'react-select/creatable';`

And it's what we don't want to use as the behaviour is not aligned with namespaces.

Reply:
thank you @konrad147 ! I will take a look!

Reply:
I was handling it by my own to be consistent with name spaces behaviour. Let's see how it looks using this `AsyncCreatable` 🤞

Reply:
Now I remember I'm not using `AsyncCreatable`, we need to keep the consistency between Folder, Namespace and Group dropdowns. And so far, in name spaces we have this:
<img width="522" alt="Screenshot 2022-09-06 at 16 43 03" src="https://user-images.githubusercontent.com/33540275/188665412-5c37b8f0-132d-48ec-818a-72025b16b962.png">

So if we want to have the same behaviour/UI we need to handle it in the same way.